### PR TITLE
Serde rfc 2822

### PR DESCRIPTION
**Pull request by chrisranderson**
> ### Thanks for contributing to chrono!
>If your feature is semver-compatible, please target the main branch;
> for semver-incompatible changes, please target the `0.5.x` branch.

I'm not sure what this means - every change is compatible with semver, right? Are you asking if it's a backwards-incompatible change?

> Please consider adding a test to ensure your bug fix/feature will not break in the future.

Done.

### CODE REVIEW

**Comment by chrisranderson** [comment]
Continuing from @pitdicker https://github.com/chronotope/chrono/issues/1292#issuecomment-1918472102:

> I could comment on a few nits, but it looks pretty much like it should.
I believe all other deserialize implementations deserialize to DateTime<FixedOffset>? We should do so here to. DateTime<Utc> discards the serialized offset info.
And all other (de)serializers support an Option variant.

I changed it to deserialize to `DateTime<FixedOffset>`. Not all other (de)serializers support an Option variant, though - the rfc3339 one doesn't, right?

File: src/datetime/serde.rs [hunk 1]
```diff
@@ -24,6 +24,10 @@ pub struct MicroSecondsTimestampVisitor;
 #[derive(Debug)]
 pub struct MilliSecondsTimestampVisitor;
 
+#[doc(hidden)]
+#[derive(Debug)]
+pub struct Rfc2822Visitor;
```
Comment (pitdicker) [inline src/datetime/serde.rs:29]: I believe you can make this private. The others are public for backwards compatibility.

File: src/datetime/serde.rs [hunk 2]
```diff
@@ -1168,6 +1172,77 @@ pub mod ts_seconds_option {
...
+        D: de::Deserializer<'de>,
+    {
+        deserializer.deserialize_str(Rfc2822Visitor)
+    }
+
+    impl<'de> de::Visitor<'de> for Rfc2822Visitor {
+        type Value = DateTime<FixedOffset>;
+
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            formatter.write_str("an RFC 2822 (email) formatted datetime string")
```
Comment (pitdicker) [inline src/datetime/serde.rs:1234]: Personally I would remove `(email) `.

File: src/datetime/serde.rs [hunk 3]
```diff
@@ -1168,6 +1172,77 @@ pub mod ts_seconds_option {
...
+    use serde::{de, ser};
+
+    /// Serialize a datetime into an RFC 2822 formatted string, e.g. "01 Jun 2016 14:31:46 -0700"
+    ///
+    /// Intended for use with `serde`s `serialize_with` attribute.
+    pub fn serialize<S>(dt: &DateTime<FixedOffset>, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: ser::Serializer,
+    {
+        serializer.serialize_str(&dt.to_rfc2822())
```
Comment (pitdicker) [inline src/datetime/serde.rs:1217]: In `format/formatting.rs` there is a `write_rfc2822` function that works without allocating. Using that will make this serializer work on `no_std` targets like the others.

You will have to adjust its `#[cfg(feature = "alloc")]` to also enable it with the `serde` feature.

**Review by pitdicker — CHANGES_REQUESTED** [review]
We are inevitably going to get a request to add an optional variant. But lets get this ready first.

**Comment by pitdicker** [comment]
Forgot to add: thank you for working on this!

And the changes are related enough to be squashed into one commit please.

**Comment by pitdicker** [comment]
@chrisranderson are you still interested in completing this PR?

**Comment by chrisranderson** [comment]
No, sorry. Gratefully, I am now employed. Sorry to leave this hanging loose.

**Comment by pitdicker** [comment]
Congratulations! And no worries.

**Comment by pitdicker** [comment]
Opened #1477.

**Pull request closed without merging**